### PR TITLE
Support conditions in JOIN ON section

### LINKED ISSUE #21868: Support filters in `JOIN ON` section
Support in `JOIN ON` section filters that affect columns from only one table. For use case see
https://stackoverflow.com/questions/5284929/mysql-left-join-with-optional-value-in-second-table

Now it's only move such conditions from `ON` section to `WHERE`, but we need to handle this filters inside join implementation. It will also fix https://github.com/ClickHouse/ClickHouse/issues/21671

### LINKED ISSUE #24437: JOIN on constant fails with 21.5.5.12 - worked with 21.4.7.3
**Describe the bug**
JOIN on constant fails with latest stable version 21.5.5.12. It worked with previous stable version 21.4.7.3.

**Does it reproduce on recent release?**
Yes. On 21.5.5.12.

**How to reproduce**
21.5.5.12

```
SELECT a.number
FROM
(
    SELECT number
    FROM numbers(5)
) AS a
INNER JOIN
(
    SELECT number
    FROM numbers(5)
) AS b ON (b.number = a.number) AND (b.number = 5)
```

**Expected behavior**
Query should run without errors.

**Error message and/or stacktrace**
```
Received exception from server (version 21.5.5):
Code: 403. DB::Exception: Received from localhost:9000. DB::Exception: Not equi-join ON expression: b.number = 5. No columns in one of equality side.: While processing b.number = 5.
```

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Support conditions in JOIN ON section. Close #21868

Detailed description / Documentation draft:

Conditions for left or right table columns can be specified in `JOIN ON` section as sequence of expression connected with `AND`, in addition to main equality. Only rows where conditions hold will be joined.

Current status:
* [x] Add more tests
* [x] Support MergeJoin 
* [x] Ensure that performance not affected, waiting for perftests in CI [_current tests passed, maybe need to write more specific?_]

### CODE REVIEW

**Comment by vdimir** [comment]
Current status:
* Need more testing
* Only HashJoin supported
* Need to ensure that performance not affected, waiting for perftests in CI

**Comment by PHaroZ** [comment]
many thanks for your work !

File: src/Interpreters/CollectJoinOnKeysVisitor.cpp [hunk 1]
```diff
@@ -71,32 +102,50 @@ void CollectJoinOnKeysMatcher::visit(const ASTFunction & func, const ASTPtr & as
...
+        }
     }
-    else if (inequality != ASOF::Inequality::None)
+
+    if (auto expr_from_table = getTableForIdentifiers(ast, false, data); expr_from_table != JoinIdentifierPos::Unknown)
     {
-        if (!data.is_asof)
-            throw Exception("JOIN ON inequalities are not supported. Unexpected '" + queryToString(ast) + "'",
-                            ErrorCodes::NOT_IMPLEMENTED);
+        data.analyzed_join.addJoinCondition(ast, isLeftIdentifier(expr_from_table));
```
Comment (abyss7) [inline src/Interpreters/CollectJoinOnKeysVisitor.cpp:129]: And this one doesn't allow functions other than `equals` to have identifiers from different sides (left/right tables). Also should add test on these functions.

File: src/Interpreters/CollectJoinOnKeysVisitor.cpp [hunk 2]
```diff
@@ -16,44 +16,75 @@ namespace ErrorCodes
...
     if (!asof_left_key || !asof_right_key)
         throw Exception("No inequality in ASOF JOIN ON section.", ErrorCodes::INVALID_JOIN_ON_EXPRESSION);
-    addJoinKeys(asof_left_key, asof_right_key, {1, 2});
+    addJoinKeys(asof_left_key, asof_right_key, {JoinIdentifierPos::Left, JoinIdentifierPos::Right});
+}
+
+void CollectJoinOnKeysMatcher::visit(const ASTIdentifier & ident, const ASTPtr & ast, CollectJoinOnKeysMatcher::Data & data)
+{
+    if (auto expr_from_table = getTableForIdentifiers(ast, false, data); expr_from_table != JoinIdentifierPos::Unknown)
+        data.analyzed_join.addJoinCondition(ast, isLeftIdentifier(expr_from_table));
```
Comment (abyss7) [inline src/Interpreters/CollectJoinOnKeysVisitor.cpp:84]: [self-note] Putting `ASTIdentifier` into `on_filter_condition_asts_*`

File: src/Interpreters/TableJoin.cpp [hunk 3]
```diff
@@ -474,4 +462,52 @@ String TableJoin::renamedRightColumnName(const String & name) const
...
+{
+    LOG_TRACE(&Poco::Logger::get("TableJoin"), "Add join condition for {} table: {}", (is_left ? "left" : "right"), queryToString(ast));
+
+    if (is_left)
+        on_filter_condition_asts_left.push_back(ast);
+    else
+        on_filter_condition_asts_right.push_back(ast);
+}
+
+ASTPtr TableJoin::buildJoinConditionColumn(const ASTs & on_filter_condition_asts)
```
Comment (abyss7) [inline src/Interpreters/TableJoin.cpp:475]: What is an expected resulting AST? I saw that we can push `ASTIdentifier` along with `ASTFunction` into `on_filter_condition_asts`.

File: src/Interpreters/TableJoin.cpp [hunk 4]
```diff
@@ -474,4 +462,52 @@ String TableJoin::renamedRightColumnName(const String & name) const
...
+}
+
+ASTPtr TableJoin::joinConditionColumn(JoinTableSide side) const
+{
+    if (side == JoinTableSide::Left)
+        return buildJoinConditionColumn(on_filter_condition_asts_left);
+    return buildJoinConditionColumn(on_filter_condition_asts_right);
+}
+
+/// Conditions for left/right table from JOIN ON section, only rows where conditions hold can be joined.
```
Comment (abyss7) [inline src/Interpreters/TableJoin.cpp:502]: "conditions are met" ?

File: src/Interpreters/join_common.cpp [hunk 5]
```diff
@@ -305,12 +319,38 @@ void checkTypesOfKeys(const Block & block_left, const Names & key_names_left, co
...
+                       const Block & block_right, const String & condition_name_right)
+{
+    auto check_cond_column_type = [](const Block & block, const String & col_name)
+    {
+        if (col_name.empty())
+            return;
+
+        DataTypePtr dtype = removeNullable(recursiveRemoveLowCardinality(block.getByName(col_name).type));
+
+        if (!dtype->equals(DataTypeUInt8{}))
```
Comment (abyss7) [inline src/Interpreters/join_common.cpp:345]: Does it mean that expression `t1.a + 5` will suffice if `a UInt8`? Should add test on this one.

File: src/Interpreters/MergeJoin.h [hunk 6]
```diff
@@ -78,6 +78,14 @@ class MergeJoin : public IJoin
     Block right_columns_to_add;
     SortedBlocksWriter::Blocks right_blocks;
 
+    Names key_names_right;
+    Names key_names_left;
+
+    /// Additional conditions for rows to join from JOIN ON section.
+    /// Only rows where condition holds can be joined.
```
Comment (abyss7) [inline src/Interpreters/MergeJoin.h:85]: "conditions are met" ?

File: src/Interpreters/HashJoin.cpp [hunk 7]
```diff
@@ -630,8 +639,33 @@ bool HashJoin::addJoinedBlock(const Block & source_block, bool check_limits)
     UInt8 save_nullmap = 0;
     if (isRightOrFull(kind) && null_map)
     {
+        /// Save rows with NULL keys
         for (size_t i = 0; !save_nullmap && i < null_map->size(); ++i)
             save_nullmap |= (*null_map)[i];
+
```
Comment (abyss7) [inline src/Interpreters/HashJoin.cpp:645]: Extra space

File: src/Interpreters/HashJoin.cpp [hunk 8]
```diff
@@ -898,9 +943,10 @@ NO_INLINE IColumn::Filter joinRightColumns(
             }
         }
 
-        auto find_result = key_getter.findKey(map, i, pool);
+        bool row_acceptable = !added_columns.isRowFiltered(i);
+        auto find_result = row_acceptable ? key_getter.findKey(map, i, pool) : decltype(key_getter.findKey(map, i, pool))();
```
Comment (abyss7) [inline src/Interpreters/HashJoin.cpp:947]: Maybe `typename KeyGetter::FindResult`?

File: src/Interpreters/HashJoin.cpp [hunk 9]
```diff
@@ -898,9 +943,10 @@ NO_INLINE IColumn::Filter joinRightColumns(
...
 
-        if (find_result.isFound())
+        if (row_acceptable && find_result.isFound())
```
Comment (abyss7) [inline src/Interpreters/HashJoin.cpp:949]: Shouldn't any `FindResult::isFound() == false` by default?

File: src/Interpreters/MergeJoin.cpp [hunk 10]
```diff
@@ -23,12 +25,50 @@ namespace ErrorCodes
...
+    extern const int ILLEGAL_COLUMN;
     extern const int LOGICAL_ERROR;
 }
 
 namespace
 {
 
+String deriveTempName(const String & name)
+{
+    return "--" + name;
```
Comment (abyss7) [inline src/Interpreters/MergeJoin.cpp:37]: Why such name?

**Review by abyss7 — CHANGES_REQUESTED** [review]
You should add some tests

File: src/Interpreters/MergeJoin.cpp (hunk 10 above)
Comment (vdimir) [reply src/Interpreters/MergeJoin.cpp:37]: Some internal table name needed, I haven't found good solution for it. This column added at the beginning of function joinBlock and removed at the end, so it's not visible outside. But we haven't some mechanism for generating temporary names that also cannot be already presented in block.
Do you have any ideas how to solve it?

File: src/Interpreters/HashJoin.cpp (hunk 9 above)
Comment (vdimir) [reply src/Interpreters/HashJoin.cpp:949]: Yes it is, checking `row_acceptable` can be removed.

File: src/Interpreters/join_common.cpp (hunk 5 above)
Comment (vdimir) [reply src/Interpreters/join_common.cpp:345]: I checked and `t1.a + 5` will be `UInt16` if `t1.a` is `UInt8`. I'll add test. Any UInt8 expression supported, like `toUInt8(t1.a + 5)`

File: src/Interpreters/TableJoin.cpp (hunk 3 above)
Comment (vdimir) [reply src/Interpreters/TableJoin.cpp:475]: It's expression containing `ASTIdentifier` or `ASTFunction` from `ON` section for one table joined with `AND`. I'll add a comment

File: src/Interpreters/CollectJoinOnKeysVisitor.cpp (hunk 1 above)
Comment (vdimir) [reply src/Interpreters/CollectJoinOnKeysVisitor.cpp:129]: Yes, `... ON ... AND t1.x > t2.x` not supported yet, I'll add test. But it works in postrgess. I think I'll try to support it later in other PR.

File: tests/queries/0_stateless/00878_join_unexpected_results.sql [hunk 11]
```diff
@@ -58,11 +59,13 @@ select '-';
...
 select '-';
-select t.*, s.* from t left join s on (s.a=t.a and t.b=s.b and t.a=toInt64(2)) order by t.a; -- {serverError 403 }
+select t.*, s.* from t left join s on (s.a=t.a and t.b=s.b and t.a=toInt64(2)) order by t.a;
 select '-';
 select t.*, s.* from t left join s on (s.a=t.a) order by t.a;
 select '-';
-select t.*, s.* from t left join s on (t.b=toInt64(2) and s.a=t.a) where s.b=2; -- {serverError 403 }
+select t.*, s.* from t left join s on (t.b=toInt64(2) and s.a=t.a) where s.b=2;
+select '-';
+select t.*, s.* from t left join s on (t.b=toInt64(2) and s.a=t.a) where t.b=2;
```
[the remark this answers is not in the document]
Comment (vdimir) [reply tests/queries/0_stateless/00878_join_unexpected_results.sql:68]: It's different in where section. But I'll change it a bit, to have not empty output.

File: src/Interpreters/CollectJoinOnKeysVisitor.cpp [hunk 12]
```diff
@@ -71,32 +102,50 @@ void CollectJoinOnKeysMatcher::visit(const ASTFunction & func, const ASTPtr & as
...
+            if (table_numbers.first == JoinIdentifierPos::Unknown)
+                throw Exception("Ambiguous column in expression '" + queryToString(ast) + "' in JOIN ON section",
+                                ErrorCodes::AMBIGUOUS_COLUMN_NAME);
+            data.analyzed_join.addJoinCondition(ast, isLeftIdentifier(table_numbers.first));
+            return;
+
+        }
+        if (table_numbers.first != JoinIdentifierPos::NotApplicable && table_numbers.second != JoinIdentifierPos::NotApplicable)
+        {
+            data.addJoinKeys(left, right, table_numbers);
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/CollectJoinOnKeysVisitor.cpp:122]: It should be supported as we support complex keys in join now. I'll add test

[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/CollectJoinOnKeysVisitor.cpp:122]: https://github.com/vdimir/ClickHouse/blob/de3198020cfa05275c9c63a16657fd2f18ec3d45/tests/queries/0_stateless/01881_join_on_conditions.sql#L57-L58

File: src/Interpreters/CollectJoinOnKeysVisitor.cpp [hunk 13]
```diff
@@ -71,32 +102,50 @@ void CollectJoinOnKeysMatcher::visit(const ASTFunction & func, const ASTPtr & as
...
+            return;
+        }
     }
-    else if (inequality != ASOF::Inequality::None)
+
+    if (auto expr_from_table = getTableForIdentifiers(ast, false, data); expr_from_table != JoinIdentifierPos::Unknown)
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/CollectJoinOnKeysVisitor.cpp:127]: It checks only condition after `;`

File: src/Interpreters/CollectJoinOnKeysVisitor.cpp (hunk 1 above)
Comment (vdimir) [reply src/Interpreters/CollectJoinOnKeysVisitor.cpp:129]: https://github.com/vdimir/ClickHouse/blob/de3198020cfa05275c9c63a16657fd2f18ec3d45/tests/queries/0_stateless/01881_join_on_conditions.sql#L52-L54

File: src/Interpreters/TableJoin.cpp (hunk 3 above)
Comment (vdimir) [reply src/Interpreters/TableJoin.cpp]: `buildJoinConditionColumn` is private static function, I added comment to TableJoin.h before related functions (addJoinCondition, joinConditionColumn) with general description
https://github.com/vdimir/ClickHouse/blob/de3198020cfa05275c9c63a16657fd2f18ec3d45/src/Interpreters/TableJoin.h#L162-L177

**Comment by vdimir** [comment]
Merged with master.

@abyss7 take a look, please, review comments are resolved

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-12064

**Pull request merged**